src/compiler/transformers/jsx.ts

```diff
@@ -24,7 +26,458 @@ namespace ts {
         }
 
         function visitorWorker(node: Node): Node {
-            return node;
+            switch (node.kind) {
+                case SyntaxKind.JsxElement:
+                    return visitJsxElement(<JsxElement>node);
+
+                case SyntaxKind.JsxSelfClosingElement:
+                    return visitJsxSelfClosingElement(<JsxSelfClosingElement>node);
+            }
+
+            Debug.fail("Unexpected node kind.");
+        }
+
+        function transformJsxChildToExpression(node: JsxChild): Expression {
+            switch (node.kind) {
+                case SyntaxKind.JsxText:
+                    return visitJsxText(<JsxText>node);
+
+                case SyntaxKind.JsxExpression:
+                    return visitJsxExpression(<JsxExpression>node);
+
+                case SyntaxKind.JsxElement:
+                    return visitJsxElement(<JsxElement>node);
+
+                case SyntaxKind.JsxSelfClosingElement:
+                    return visitJsxSelfClosingElement(<JsxSelfClosingElement>node);
+            }
+
+            Debug.fail("Unexpected node kind.");
+        }
+
+        function visitJsxElement(node: JsxElement) {
+            return visitJsxOpeningLikeElement(node.openingElement, node.children);
+        }
+
+        function visitJsxSelfClosingElement(node: JsxSelfClosingElement) {
+            return visitJsxOpeningLikeElement(node, /*children*/ undefined);
+        }
+
+        function visitJsxOpeningLikeElement(node: JsxOpeningLikeElement, children: JsxChild[]) {
+            // We must the node onto the node stack if it is not already at the top.
+            const tagName = getTagName(node);
+            let objectProperties: Expression;
+            if (node.attributes.length === 0) {
+                // When there are no attributes, React wants "null"
+                objectProperties = createNull();
+            }
+            else {
+                // Either emit one big object literal (no spread attribs), or
+                // a call to React.__spread
+                const attrs = node.attributes;
+                if (forEach(attrs, isJsxSpreadAttribute)) {
+                    const segments: Expression[] = [];
+                    let properties: ObjectLiteralElement[] = [];

```

Comment:
We must always emit *at least one* object literal. I'll change this to make this more explicit.